server/src/main/kotlin/com/trib3/server/graphql/GraphQLWebSocketSubscriber.kt

```diff
@@ -0,0 +1,289 @@
+package com.trib3.server.graphql
+
+import com.trib3.server.filters.RequestIdFilter
+import graphql.ExecutionInput
+import graphql.ExecutionResult
+import graphql.GraphQL
+import io.reactivex.Flowable
+import io.reactivex.Scheduler
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import mu.KotlinLogging
+import org.eclipse.jetty.websocket.api.StatusCode
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+import org.reactivestreams.Subscription
+import java.util.concurrent.TimeUnit
+import javax.annotation.Nullable
+
+private val log = KotlinLogging.logger {}
+
+/**
+ * rx [Subscriber] that listens for events on coming from the WebSocket managed
+ * by a [GraphQLWebSocketAdapter], and implements the apollo graphql-ws protocol
+ * from https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md
+ *
+ * Handling some WebSocket events results in new rx subscriptions (eg, starting a query or
+ * a keepalive timer).  The handlers for those subscriptions must inject back into the
+ * original subscription via [adapter.emitter] if they need to modify the subscriber's state
+ * or send data to the WebSocket client.
+ */
+class GraphQLWebSocketSubscriber(
+    @Nullable val graphQL: GraphQL?,
+    val graphQLConfig: GraphQLConfig,
+    val scheduler: Scheduler = Schedulers.io(), // default to io scheduler for actual work
+    val keepAliveScheduler: Scheduler = Schedulers.computation() // default to computation for the KA interval
+) : Subscriber<OperationMessage<*>>, Disposable {
+    internal lateinit var adapter: GraphQLWebSocketAdapter
+    private lateinit var subscription: Subscription
+    private val disposables = CompositeDisposable() // all downstream subscriptions stored here for disposal
+    private var keepAlive: Disposable? = null
+    private val queries = mutableMapOf<String, Disposable>()
+
+    /**
+     * Store the subscription so we can start requesting events from it later
+     */
+    override fun onSubscribe(s: Subscription) {
+        subscription = s
```

Comment:
possible race conditions with setting subscription?

Reply:
kind of -- the rxjava guarantees around the order of things i _think_ make it so that we're ok from a race condition perspective.  HOWEVER, they provide a `ResourceSubscriber` class we should extend instead of doing the subscription management and `disposables` management all ourself. will update to that.